cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AbstractNotificationSubscriberService.java

```diff
@@ -107,97 +105,148 @@ protected void shutDown() {
-            Message message = iterator.next();
-            lastFetchedMessageId = message.getId();
-            try {
-              Notification notification = GSON.fromJson(new String(message.getPayload(), StandardCharsets.UTF_8),
-                                                        Notification.class);
-              fetchedNotifications.add(notification);
-            } catch (JsonSyntaxException e) {
-              LOG.warn("Failed to decode message with id {}. Skipped. ", message.getId(), e);
-              // lastFetchedMessageId was updated at the beginning of the loop to skip this message in next fetch
-            }
-          }
-        }
+        // Collects batch of messages for processing.
+        // The fetch may be transactional, and it's ok to have the fetching and the processing happen in two
+        // non-overlapping transactions, as long as the processing transaction starts after the fetching one.
+        Stopwatch stopwatch = new Stopwatch().start();
+        final List<Message> messages = fetchMessages(messageId);
+        metricsContext.gauge("tms.fetch.time.ms", stopwatch.elapsedTime(TimeUnit.MILLISECONDS));
+        metricsContext.increment("tms.fetch.messages", messages.size());
 
-        // Sleep for configured number of milliseconds if there are no notifications
-        if (lastFetchedMessageId == null) {
-          return cConf.getLong(Constants.Scheduler.EVENT_POLL_DELAY_MILLIS);
+        // Return if stopping or request to sleep for configured number of milliseconds if there are no notifications
+        if (stopping || messages.isEmpty()) {
+          return emptyFetchDelayMillis;
         }
 
-        final String lastMessageId = lastFetchedMessageId;
-        // Execute processed notifications in one transaction
-        Retries.supplyWithRetries(new Supplier<Void>() {
+        stopwatch.reset().start();
+        // Process the notifications and record the message id of where the processing is up to.
+        messageId = Transactionals.execute(transactional, new TxCallable<String>() {
           @Override
-          public Void get() {
-            Transactions.executeUnchecked(transactional, new TxCallable<Void>() {
-              @Override
-              public Void call(DatasetContext context) throws Exception {
-                processFetchedNotifications(context, fetchedNotifications);
-                updateMessageId(context, lastMessageId);
-                return null;
-              }
-            });
-            return null;
+          public String call(DatasetContext context) throws Exception {
+            NotificationIterator iterator = new NotificationIterator(topic, messages.iterator());
+            processNotifications(context, iterator);
+            String lastMessageId = iterator.getLastMessageId();
+
+            // If the processNotifications hasn't consumed anything, there is nothing to save.
+            if (lastMessageId == null) {
```

Comment:
That kind of making the contract unclear. The current contract of the `NotificationIterator.getLastMessageId` is to return the last message id of the message being returned by the last call to `next()`. It's better to keep that simple.